gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -56,15 +60,21 @@
  * </ul>
  */
 @InternalApi
-public class Watchdog implements Runnable {
+public class Watchdog implements Runnable, BackgroundResource {
   // Dummy value to convert the ConcurrentHashMap into a Set
   private static Object PRESENT = new Object();
   private final ConcurrentHashMap<WatchdogStream, Object> openStreams = new ConcurrentHashMap<>();
 
   private final ApiClock clock;
+  private final ScheduledExecutorService executor;
+  private final ScheduledFuture<?> future;
 
-  public Watchdog(ApiClock clock) {
+  public Watchdog(ApiClock clock, Duration scheduleInterval, ScheduledExecutorService executor) {
     this.clock = Preconditions.checkNotNull(clock, "clock can't be null");
+    this.executor = executor;
+    this.future =
+        executor.scheduleAtFixedRate(
+            this, scheduleInterval.toMillis(), scheduleInterval.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
Please avoid having constructors with side effects. I think it would be better to move the scheduling to a start method and call that from either the provider or a static factory method

Reply:
Thanks for the suggestion. Added Watchdog.create and shifted the scheduling into start().